services/service.go

```diff
@@ -29,97 +34,95 @@ type Test struct {
 
 //Service represents the API to test
 type Service struct {
-	ID       string    // unique random ID assinged at run-time
-	Name     string    `json:"name"`
-	URL      string    `json:"url"`
-	Disabled bool      `json:"disabled"`
-	Tests    Test      `json:"tests"`
-	Alerters []Alerter `json:"alerters"`
+	Name     string              `json:"name"`
+	URL      string              `json:"url"`
+	Disabled bool                `json:"disabled"`
+	Failed   int                 // 0 nil, 1 yes, 2 no
+	Tests    Test                `json:"tests"`
+	Handlers []*alerting.Alerter `json:"alerters"`
 	Logger   zap.Logger
 }
 
-//RunAll performs all tests for the given service
-func (s *Service) RunAll() error {
-	if s.Disabled {
-		return nil
-	}
-
-	if err := s.Ping(); err != nil {
-		return err
-	}
-
-	if err := s.JSON(); err != nil {
-		return err
-	}
-
-	if err := s.Time(); err != nil {
-		return err
+//checks if there is Internet connection by pinging Google
+func hasInternet(logger zap.Logger) bool {
+	notify := func(err error, duration time.Duration) {
+		logger.Warn(
+			"Encountered error during run",
+			zap.Int64("duration", duration.Nanoseconds()/int64(time.Millisecond)),
+			zap.Error(err),
+		)
 	}
 
-	return nil
-}
-
-//Ping checks if a service is available
-func (s *Service) Ping() error {
-
-	// Make request
 	operation := func() error {
-		response, err := http.Get(s.URL)
+		response, err := http.Get("https://www.google.com")
 
 		if err != nil {
+			// Pass this error back, the notify handler will log it for us
 			return err
 		}
 
 		defer response.Body.Close()
 		io.Copy(ioutil.Discard, response.Body)
 
-		if response.StatusCode != http.StatusOK {
-			return fmt.Errorf("Service HTTP status code is %d, %d expected", response.StatusCode, http.StatusOK)
+		if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
+			// Pass this error back, the notify handler will log it for us
+			return errors.New("Returned status code was not acceptable")
 		}
 		return err
 	}
 
-	return retry.Retrying(operation, s.Logger)
+	err := backoff.RetryNotify(operation, backoff.NewExponentialBackOff(), notify)
+	return err == nil
 }
 
-//JSON test if the service response is valid json
-func (s *Service) JSON() error {
-
-	operation := func() error {
-		response, err := http.Get(s.URL)
-		if err != nil {
-			return errors.New("Service not reachable")
+//PerformChecks calls RunAll and handles calling the alerters as needed
+func (s *Service) PerformChecks() {
+	if err := s.RunAll(); err != nil {
+		// Sanity check, do we have an Internet connection?
+		if !hasInternet(s.Logger) {
+			a := alerting.CreateAlert("Peekaboo has no Internet connection.", s.Logger)
```

Comment:
Yes, I actually had a comment to that effect earlier, but left it out.

It does need to be more than a log message though, since we want the timestamp field from it, but i have removed the `logger` field and the `Log` func from this type now. So it should be much cleaner, I hope.